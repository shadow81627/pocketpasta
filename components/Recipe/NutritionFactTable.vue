<template>
  <b-card
    no-body
    tag="section"
    class="performance-facts"
    itemscope
    itemtype="http://schema.org/NutritionInformation"
  >
    <b-card-header header-tag="header" class="p-1">
      <b-button
        v-b-toggle:collapse-performance-facts
        block
        variant="none"
        squared
        class="text-left"
      >
        <h2>Nutrition Facts</h2>
      </b-button>
    </b-card-header>
    <b-collapse id="collapse-performance-facts" visible>
      <b-card-body>
        <header class="performance-facts__header">
          <p itemprop="servingSize">Serving Size: {{ servingSize }}</p>
          <!-- <p>Serving Per Container 8</p> -->
        </header>
        <table class="performance-facts__table">
          <thead>
            <tr>
              <th colspan="3" class="small-info">
                Amount Per Serving
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th colspan="2">
                <b>Calories</b>
                <span itemprop="calories">{{ calories }}</span>
              </th>
              <td>
                Calories from Fat
                <span itemprop="fiberContent">{{ fiberContent }}</span>
              </td>
            </tr>
            <tr class="thick-row">
              <td colspan="3" class="small-info">
                <b>% Daily Value*</b>
              </td>
            </tr>
            <tr>
              <th colspan="2">
                <b>Total Fat</b>
                <span itemprop="fatContent">{{ fatContent }}</span>
              </th>
              <td>
                <!-- <b>22%</b> -->
              </td>
            </tr>
            <tr>
              <td class="blank-cell" />
              <th>
                Saturated Fat
                <span itemprop="saturatedFatContent">{{
                  saturatedFatContent
                }}</span>
              </th>
              <td>
                <!-- <b>22%</b> -->
              </td>
            </tr>
            <tr>
              <td class="blank-cell" />
              <th>
                Trans Fat
                <span itemprop="unsaturatedFatContent">{{
                  unsaturatedFatContent
                }}</span>
              </th>
              <td />
            </tr>
            <tr>
              <th colspan="2">
                <b>Cholesterol</b>
                <span itemprop="cholesterolContent">{{
                  cholesterolContent
                }}</span>
              </th>
              <td>
                <!-- <b>18%</b> -->
              </td>
            </tr>
            <tr>
              <th colspan="2">
                <b>Sodium</b>
                <span itemprop="sodiumContent">{{ sodiumContent }}</span>
              </th>
              <td>
                <!-- <b>2%</b> -->
              </td>
            </tr>
            <tr>
              <th colspan="2">
                <b>Total Carbohydrate</b>
                <span itemprop="carbohydrateContent">{{
                  carbohydrateContent
                }}</span>
              </th>
              <td>
                <!-- <b>6%</b> -->
              </td>
            </tr>
            <tr>
              <td class="blank-cell" />
              <th>
                Dietary Fiber
                <span itemprop="fiberContent">{{ fiberContent }}</span>
              </th>
              <td>
                <!-- <b>4%</b> -->
              </td>
            </tr>
            <tr>
              <td class="blank-cell" />
              <th>
                Sugars
                <span itemprop="sugarContent">{{ sugarContent }}</span>
              </th>
              <td />
            </tr>
            <tr class="thick-end">
              <th colspan="2">
                <b>Protein</b>
                <span itemprop="proteinContent">{{ proteinContent }}</span>
              </th>
              <td />
            </tr>
          </tbody>
        </table>

        <!-- <table class="performance-facts__table--grid">
      <tbody>
        <tr>
          <td colspan="2">
            Vitamin A 10%
          </td>
          <td>
            Vitamin C 0%
          </td>
        </tr>
        <tr class="thin-end">
          <td colspan="2">
            Calcium 10%
          </td>
          <td>
            Iron 6%
          </td>
        </tr>
      </tbody>
    </table> -->

        <p class="small-info">
          * Percent Daily Values are based on a 2,000 calorie diet. Your daily
          values may be higher or lower depending on your calorie needs:
        </p>

        <!-- <table class="performance-facts__table--small small-info">
      <thead>
        <tr>
          <td colspan="2" />
          <th>Calories:</th>
          <th>2,000</th>
          <th>2,500</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th colspan="2">Total Fat</th>
          <td>Less than</td>
          <td>65g</td>
          <td>80g</td>
        </tr>
        <tr>
          <td class="blank-cell" />
          <th>Saturated Fat</th>
          <td>Less than</td>
          <td>20g</td>
          <td>25g</td>
        </tr>
        <tr>
          <th colspan="2">Cholesterol</th>
          <td>Less than</td>
          <td>300mg</td>
          <td>300 mg</td>
        </tr>
        <tr>
          <th colspan="2">Sodium</th>
          <td>Less than</td>
          <td>2,400mg</td>
          <td>2,400mg</td>
        </tr>
        <tr>
          <th colspan="3">Total Carbohydrate</th>
          <td>300g</td>
          <td>375g</td>
        </tr>
        <tr>
          <td class="blank-cell" />
          <th colspan="2">Dietary Fiber</th>
          <td>25g</td>
          <td>30g</td>
        </tr>
      </tbody>
    </table> -->

        <!-- <p class="small-info">
      Calories per gram:
    </p>
    <p class="small-info text-center">
      Fat 9 &bull; Carbohydrate 4 &bull; Protein 4
    </p> -->
      </b-card-body>
    </b-collapse>
  </b-card>
</template>

<script>
import { BCollapse, VBToggle } from 'bootstrap-vue';
export default {
  components: {
    BCollapse,
  },
  directives: { 'b-toggle': VBToggle },
  inheritAttrs: false,
  props: {
    calories: { type: String, default: null },
    carbohydrateContent: { type: String, default: null },
    cholesterolContent: { type: String, default: null },
    fatContent: { type: String, default: null },
    fiberContent: { type: String, default: null },
    proteinContent: { type: String, default: null },
    saturatedFatContent: { type: String, default: null },
    servingSize: { type: String, default: null },
    sodiumContent: { type: String, default: null },
    sugarContent: { type: String, default: null },
    unsaturatedFatContent: { type: String, default: null },
  },
};
</script>

<style lang="scss" scoped>
.performance-facts {
  // border: 1px solid black;
  // margin: 20px;
  // float: left;
  // width: 280px;
  // padding: 0.5rem;
  table {
    border-collapse: collapse;
  }
}
.performance-facts__title {
  font-weight: bold;
  font-size: 2rem;
  margin: 0 0 0.25rem 0;
}
.performance-facts__header {
  border-bottom: 10px solid black;
  padding: 0 0 0.25rem 0;
  margin: 0 0 0.5rem 0;
  p {
    margin: 0;
  }
}
.performance-facts__table {
  width: 100%;
  thead tr {
    th,
    td {
      border: 0;
    }
  }
  th,
  td {
    font-weight: normal;
    text-align: left;
    padding: 0.25rem 0;
    border-top: 1px solid black;
    white-space: nowrap;
  }
  td {
    &:last-child {
      text-align: right;
    }
  }
  .blank-cell {
    width: 1rem;
    border-top: 0;
  }
  .thick-row {
    th,
    td {
      border-top-width: 5px;
    }
  }
}
.small-info {
  font-size: 0.7rem;
}

.performance-facts__table--small {
  @extend .performance-facts__table;
  border-bottom: 1px solid #999;
  margin: 0 0 0.5rem 0;
  thead {
    tr {
      border-bottom: 1px solid black;
    }
  }
  td {
    &:last-child {
      text-align: left;
    }
  }
  th,
  td {
    border: 0;
    padding: 0;
  }
}

.performance-facts__table--grid {
  @extend .performance-facts__table;
  margin: 0 0 0.5rem 0;
  td {
    &:last-child {
      text-align: left;
      &::before {
        content: 'â€¢';
        font-weight: bold;
        margin: 0 0.25rem 0 0;
      }
    }
  }
}

.text-center {
  text-align: center;
}
.thick-end {
  border-bottom: 10px solid black;
}
.thin-end {
  border-bottom: 1px solid black;
}
</style>
