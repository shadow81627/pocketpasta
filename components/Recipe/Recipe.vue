<template>
  <div class="recipe">
    <h1>{{ recipe.name }}</h1>
    <p>{{ recipe.description }}</p>

    <vue-plyr
      v-if="
        recipe.video &&
        recipe.video.embedUrl &&
        recipe.video.embedUrl.includes('youtube')
      "
    >
      <div
        data-plyr-provider="youtube"
        :data-plyr-embed-id="recipe.video.embedUrl"
        style="padding-top: 56.25%;"
      />
    </vue-plyr>
    <div v-else-if="imageData && imageData.url">
      <b-img-lazy
        :src="imageData.url"
        class="img-fluid mx-auto d-block"
        :alt="recipe.name"
        throttle="100"
        itemprop="image"
        :width="640"
        :height="360"
      />
    </div>
    <!-- <p>Author: {{ recipe.author }}</p> -->
    <!-- <p>Published: {{ recipe.datePublished }}</p> -->

    <share class="py-4" />

    <p>
      <number-text :text="recipe.recipeYield" />
    </p>
    <b-row>
      <b-col md="6">
        <b-card
          v-if="recipe.recipeIngredient"
          no-body
          tag="section"
          class="mb-4"
        >
          <b-card-header header-tag="header" class="p-1 text-left" role="tab">
            <b-button
              v-b-toggle:collapse-ingredient
              block
              variant="none"
              squared
              class="text-left"
              ><h2>Ingredients</h2></b-button
            >
          </b-card-header>
          <b-collapse id="collapse-ingredient" visible>
            <b-card-body>
              <ol class="list-group-flush pl-0 mb-0">
                <li
                  v-for="ingredient in recipe.recipeIngredient"
                  :key="ingredient"
                  class="list-group-item"
                >
                  <fraction-text :text="ingredient" />
                </li>
              </ol>
            </b-card-body>
          </b-collapse>
        </b-card>
      </b-col>
      <b-col md="6">
        <b-card
          v-if="recipe.recipeInstructions"
          no-body
          tag="section"
          class="mb-4"
        >
          <b-card-header header-tag="header" class="p-1" role="tab">
            <b-button
              v-b-toggle:collapse-instructions
              block
              variant="none"
              squared
              class="text-left"
              ><h2>Instructions</h2></b-button
            >
          </b-card-header>
          <b-collapse id="collapse-instructions" visible>
            <b-card-body>
              <ol
                v-if="Array.isArray(recipe.recipeInstructions)"
                class="list-group-flush pl-0 mb-0"
              >
                <li
                  v-for="instruction in recipe.recipeInstructions"
                  :key="instruction.text"
                  class="list-group-item"
                >
                  <fraction-text :text="instruction.text" />
                </li>
              </ol>
              <p v-else>{{ recipe.recipeInstructions }}</p>
            </b-card-body>
          </b-collapse>
        </b-card>
      </b-col>
    </b-row>

    <nutrition-fact-table
      v-if="recipe.nutrition"
      v-bind="recipe.nutrition"
      class="my-4"
    />

    <section class="mb-2">
      <h2 class="h4">Tags</h2>
      <div class="list-group-flush">
        <keywords
          :tags="recipe.keywords ? recipe.keywords.split(',') : []"
          :label="null"
          class="list-group-item"
        />
      </div>
    </section>

    <section v-if="recipe.sameAs">
      <h2 class="h4">References</h2>
      <div class="list-group-flush">
        <span
          v-for="reference in recipe.sameAs"
          :key="reference"
          class="list-group-item"
        >
          <a :href="reference" target="_blank" itemprop="url" rel="noopener">{{
            reference
          }}</a>
        </span>
      </div>
    </section>
  </div>
</template>

<script>
import NumberText from '@/components/text/NumberText';
import FractionText from '@/components/FractionText';
import Keywords from '@/components/Keywords';
import Share from '@/components/Social/Share';
import NutritionFactTable from '@/components/Recipe/NutritionFactTable';
import { BCollapse, VBToggle, BRow, BCol } from 'bootstrap-vue';
import VuePlyr from 'vue-plyr/dist/vue-plyr.ssr.js';
import 'plyr/dist/plyr.css';
export default {
  components: {
    NumberText,
    FractionText,
    Keywords,
    Share,
    NutritionFactTable,
    BCollapse,
    BRow,
    BCol,
    VuePlyr,
  },
  directives: { 'b-toggle': VBToggle },
  inheritAttrs: false,
  computed: {
    recipe() {
      // parse id param to int for id lookup
      const id = parseInt(this.$route.params.id, 10);
      const recipe = JSON.parse(
        JSON.stringify(this.$store.getters.getRecipeById(id)),
      );
      recipe.recipeIngredient = [...new Set(recipe.recipeIngredient)];

      return recipe;
    },
    imageData() {
      function cloudinaryify(image) {
        if (!image.startsWith('https://res.cloudinary.com')) {
          return `https://res.cloudinary.com/pocketpasta/image/fetch/w_640,h_360,ar_16:9,c_fill,f_auto,q_auto/${image}`;
        } else {
          return image;
        }
      }
      if (this.recipe.image) {
        if (
          typeof this.recipe.image === 'object' &&
          this.recipe.image !== null &&
          !Array.isArray(this.recipe.image)
        ) {
          const { url, width, height } = this.recipe.image;
          return { url: cloudinaryify(url), width, height };
        } else if (Array.isArray(this.recipe.image)) {
          return { url: cloudinaryify(this.recipe.image[0]) };
        } else {
          return { url: cloudinaryify(this.recipe.image) };
        }
      } else {
        return null;
      }
    },
  },

  head() {
    return {
      title: this.recipe.name,
      meta: [
        {
          hid: 'description',
          name: 'description',
          content: this.recipe.description,
        },
        {
          hid: 'og:description',
          name: 'og:description',
          content: this.recipe.description,
        },
      ],
      script: [
        {
          json: this.recipe,
          type: 'application/ld+json',
        },
      ],
    };
  },
};
</script>
