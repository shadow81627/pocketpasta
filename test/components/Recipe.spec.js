import Recipe from '@/components/Recipe/Recipe';
import recipe from '@/assets/link-data/recipes/spaghetti.json';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import BootstrapVuePlugin from 'bootstrap-vue';
import VueMeta from 'vue-meta';
// import VueRouter from 'vue-router';

const localVue = createLocalVue();

localVue.use(Vuex);
// localVue.use(VueRouter);
localVue.use(BootstrapVuePlugin);
localVue.use(VueMeta, {
  keyName: 'head',
  attribute: 'data-vue-meta',
  ssrAttribute: 'data-vue-meta-server-rendered',
  tagIDKeyName: 'hid',
  refreshOnceOnNavigation: true,
});

// const router = new VueRouter();

const $route = {
  path: '/recipes/detail/1',
  params: { id: '1' },
  fullPath: '/recipes/detail/1',
  name: 'recipes-detail-id',
};

describe('Recipe', () => {
  let getters;
  let state;
  let store;
  let mutations;

  beforeEach(() => {
    state = {
      recipes: [{ ...recipe, id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }],
    };

    getters = {
      getRecipeById: (state) => (id) =>
        state.recipes.find((recipe) => recipe.id === id),
    };

    mutations = {
      setTheme(currentState, theme) {
        currentState.theme = theme;
      },
    };

    store = new Vuex.Store({
      state,
      getters,
      mutations,
    });
  });

  const factory = () =>
    shallowMount(Recipe, {
      store,
      // router,
      localVue,
      mocks: {
        $route,
      },
    });

  test('mounts properly', () => {
    const wrapper = factory();
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test('renders properly', () => {
    const wrapper = factory();
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('head', () => {
    const wrapper = factory();
    expect(wrapper.vm.$metaInfo).toBeTruthy();
  });
});
