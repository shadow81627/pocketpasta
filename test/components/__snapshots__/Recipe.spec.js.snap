// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe renders properly 1`] = `
"<div class=\\"recipe\\">
  <h1 itemprop=\\"name\\"></h1>
  <p itemprop=\\"description\\">
    <read-more-stub morestr=\\"Read more\\" lessstr=\\"Read less\\" link=\\"#\\" stop=\\"120\\" clamp=\\"...\\"></read-more-stub>
  </p>
  <!---->
  <section>
    <v-row-stub tag=\\"div\\">
      <v-col-stub tag=\\"div\\">
        <v-chip-stub activeclass=\\"\\" tag=\\"span\\" active=\\"true\\" closeicon=\\"$delete\\" filtericon=\\"$complete\\" label=\\"true\\" style=\\"background: none;\\"><span itemprop=\\"ratingValue\\">4.9</span>
          <v-rating-stub color=\\"primary\\" opendelay=\\"0\\" closedelay=\\"0\\" ripple=\\"true\\" backgroundcolor=\\"accent\\" dense=\\"true\\" emptyicon=\\"$ratingEmpty\\" fullicon=\\"$ratingFull\\" halficon=\\"$ratingHalf\\" length=\\"5\\" readonly=\\"true\\" value=\\"4.93\\"></v-rating-stub> <span itemprop=\\"ratingCount\\">(1)</span></v-chip-stub>
        <!---->
        <!---->
        <v-chip-stub activeclass=\\"\\" tag=\\"span\\" active=\\"true\\" closeicon=\\"$delete\\" filtericon=\\"$complete\\" label=\\"true\\" style=\\"background: none; overflow: visible; padding: 0px;\\">
          <share-stub class=\\"d-inline-block\\"></share-stub>
        </v-chip-stub>
      </v-col-stub>
    </v-row-stub>
  </section>
  <v-expansion-panels-stub value=\\"0,1,2\\" activeclass=\\"v-item--active\\" multiple=\\"true\\" accordion=\\"true\\">
    <v-expansion-panel-stub>
      <v-expansion-panel-header-stub expandicon=\\"$expand\\">
        <h2>Ingredients</h2>
      </v-expansion-panel-header-stub>
      <v-expansion-panel-content-stub eager=\\"true\\">
        <v-data-iterator-stub mobilebreakpoint=\\"600\\" items=\\"\\" options=\\"[object Object]\\" sortby=\\"\\" sortdesc=\\"\\" customsort=\\"function sortItems(items, sortBy, sortDesc, locale, customSorters) {
          if (sortBy === null || !sortBy.length) return items;
          var stringCollator = new Intl.Collator(locale, {
            sensitivity: 'accent',
            usage: 'sort'
          });
          return items.sort(function (a, b) {
            var _a, _b;

            for (var i = 0; i < sortBy.length; i++) {
              var sortKey = sortBy[i];
              var sortA = getObjectValueByPath(a, sortKey);
              var sortB = getObjectValueByPath(b, sortKey);

              if (sortDesc[i]) {
                _a = __read([sortB, sortA], 2), sortA = _a[0], sortB = _a[1];
              }

              if (customSorters &amp;&amp; customSorters[sortKey]) {
                var customResult = customSorters[sortKey](sortA, sortB);
                if (!customResult) continue;
                return customResult;
              } // Check if both cannot be evaluated


              if (sortA === null &amp;&amp; sortB === null) {
                continue;
              }

              _b = __read([sortA, sortB].map(function (s) {
                return (s || '').toString().toLocaleLowerCase();
              }), 2), sortA = _b[0], sortB = _b[1];

              if (sortA !== sortB) {
                if (!isNaN(sortA) &amp;&amp; !isNaN(sortB)) return Number(sortA) - Number(sortB);
                return stringCollator.compare(sortA, sortB);
              }
            }

            return 0;
          });
        }\\" page=\\"1\\" itemsperpage=\\"-1\\" groupby=\\"\\" groupdesc=\\"\\" customgroup=\\"function groupItems(items, groupBy, groupDesc) {
          var key = groupBy[0];
          var groups = [];
          var current = null;

          for (var i = 0; i < items.length; i++) {
            var item = items[i];
            var val = getObjectValueByPath(item, key);

            if (current !== val) {
              current = val;
              groups.push({
                name: val,
                items: []
              });
            }

            groups[groups.length - 1].items.push(item);
          }

          return groups;
        }\\" locale=\\"en-US\\" customfilter=\\"function searchItems(items, search) {
          if (!search) return items;
          search = search.toString().toLowerCase();
          if (search.trim() === '') return items;
          return items.filter(function (item) {
            return Object.keys(item).some(function (key) {
              return defaultFilter(getObjectValueByPath(item, key), search, item);
            });
          });
        }\\" serveritemslength=\\"-1\\" itemkey=\\"id\\" value=\\"\\" expanded=\\"\\" noresultstext=\\"$vuetify.dataIterator.noResultsText\\" nodatatext=\\"$vuetify.noDataText\\" loadingtext=\\"$vuetify.dataIterator.loadingText\\" hidedefaultfooter=\\"true\\" selectablekey=\\"isSelectable\\"></v-data-iterator-stub>
      </v-expansion-panel-content-stub>
    </v-expansion-panel-stub>
    <v-expansion-panel-stub class=\\"my-4\\">
      <v-expansion-panel-header-stub expandicon=\\"$expand\\">
        <h2>Instructions</h2>
      </v-expansion-panel-header-stub>
      <v-expansion-panel-content-stub eager=\\"true\\">
        <p></p>
      </v-expansion-panel-content-stub>
    </v-expansion-panel-stub>
    <!---->
    <!---->
    <!---->
  </v-expansion-panels-stub>
</div>"
`;
